src/ucp/wireup/select.c

```diff
@@ -1254,45 +1318,97 @@ ucs_status_t ucp_wireup_select_lanes(ucp_ep_h ep, const ucp_ep_params_t *params,
     ucp_worker_h worker   = ep->worker;
     ucp_context_h context = worker->context;
     double am_score       = 0.0;
+    /* The flags below indicates whether we need to use RMA/AMO
+     * emulation over AM or not */
+    int rma_need_am       = 0;
+    int amo_need_am       = 0;
+    /* The flag indicates should add AM lane in case of RMA/AMO lanes
+     * better fits (e.g., neither AM nor RMA/AMO provides enough maximum
+     * number of EPs supported on ifaces, but RMA/AMO are closer to the
+     * ideal value) */
+    int add_am_lane       = 1;
+    size_t rma_best_max_num_eps;
+    size_t amo_best_max_num_eps;
+    size_t am_best_max_num_eps;
     ucp_wireup_lane_desc_t lane_descs[UCP_MAX_LANES];
     ucp_rsc_index_t rsc_index;
     ucp_md_index_t md_index;
     ucp_lane_index_t lane;
     ucp_lane_index_t i;
     ucs_status_t status;
-    int need_am = 0;
 
     memset(lane_descs, 0, sizeof(lane_descs));
 
     status = ucp_wireup_add_rma_lanes(ep, params, ep_init_flags, address_count,
                                       address_list, lane_descs, &key->num_lanes,
-                                      &need_am);
+                                      1, &rma_best_max_num_eps, &rma_need_am);
```

Comment:
maybe at first try to select AM lane to obtain maximum possible `max_eps` (without adding it).
Then pass this value to other lane selections. Probably this would simplify the flow